### ISSUE TITLE
Discrepancy of matrix multiplication due to the size

### ISSUE THREAD

**Issue by tstumm** [pytorch/pytorch#34060]
## 🐛 Bug

`MultiHeadAttention` seems to be influenced by the batch size. Calculating attention for some random tensor of shape `(L, N, E)` and with a slice `[:, 0, :]` with unsqueezed dim=1, yields different results for the slice `[:, 0, :]`.

## To Reproduce

Code to reproduce the behaviour:

```python
import math

import torch
from torch.nn.modules.activation import MultiheadAttention

torch.manual_seed(42)
data = torch.randn(3, 2, 128) * math.sqrt(128)
data_0 = data[:, 0].unsqueeze(1)
att = MultiheadAttention(128, 8, 0.)
assert torch.allclose(att.forward(data, data, data)[0][:, 0], 
                      att.forward(data_0, data_0, data_0)[0][:, 0])
```

## Expected behavior

Calculating attention using `MultiheadAttention` for some specific instance should return the same result, no matter the number of instances in the batch.
This implies
```python
torch.allclose(att.forward(data, data, data)[0][:, 0], att.forward(data_0, data_0, data_0)[0][:, 0]) == True
```

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: GPU 0: Quadro P2000
Nvidia driver version: 418.96
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] pytorch-lightning==0.6.0
[pip] torch==1.4.0
[pip] torchvision==0.5.0
[conda] libblas                   3.8.0                    14_mkl    conda-forge
[conda] libcblas                  3.8.0                    14_mkl    conda-forge
[conda] liblapack                 3.8.0                    14_mkl    conda-forge
[conda] mkl                       2019.4                      245
[conda] pytorch                   1.4.0           py3.7_cuda101_cudnn7_0    pytorch
[conda] pytorch-lightning         0.6.0                    pypi_0    pypi
[conda] torchvision               0.5.0                py37_cu101    pytorch

## Additional context

Does not work on neither CPU nor GPU.

cc @ezyang @gchanan @zou3519

**Comment by tstumm** [comment]
I could pinpoint the issue to the linear transformations for q,k and v. So, this seems to be caused by `torch.matmul`.

```python
import math

import torch
from torch.nn import functional as F
from torch.nn.modules.activation import MultiheadAttention

torch.manual_seed(42)
data = torch.randn(3, 2, 128) * math.sqrt(128)
data_0 = data[:, 0].unsqueeze(1)
att = MultiheadAttention(128, 8, 0.)
att.eval()

batched = F.linear(data, att.in_proj_weight, att.in_proj_bias)
single = F.linear(data_0, att.in_proj_weight, att.in_proj_bias)
assert torch.allclose(batched[:, 0], single[:, 0])
```

**Comment by ezyang** [comment]
cc @zhangguanheng66 marking as high priority as the results are silently different.

**Comment by tstumm** [comment]
The error is neglible if double precision is used, so most probably a floating point issue.
Nevertheless, F.linear / torch.matmul needs to be numerically more robust in the FP32 case.

**Comment by zhangguanheng66** [comment]
@tstumm is correct. This is expected and not a pytorch specific thing. A float number will only give you 6/7 digits precision, which is true in the above case. If you want to double the precision, you need to use `double()` or `float64`, like
```
batched = F.linear(data.double(), att.in_proj_weight.double(), att.in_proj_bias.double())
single = F.linear(data_0.double(), att.in_proj_weight.double(), att.in_proj_bias.double())
torch.allclose(batched[:, 0], single[:, 0])
```

Feel free to re-open the issue if you have any follow-up questions.

**Comment by ezyang** [comment]
@zhangguanheng66 I don't think float versus double precision has anything to do with the issue here. What is likely happening is the size of the matrix multiply is causing our internal algorithm to switch between two different versions, which have different numerical properties. I agree this is surprising to users, although there may not be too much we can do about it (maybe #15359 can help). Should we keep this open and just deprioritize?

**Comment by zhangguanheng66** [comment]
Sure. We can keep the issue open to keep the snippet code.

**Comment by colesbury** [comment]
#15359 will not change the behavior. Bitwise determinism is only a goal for operations with the same inputs. The distributive property does not hold for floating point calculations (nor does associativity) and we shouldn't expect it to hold for larger ops built upon them. That's surprising, but I don't see much value in keeping these sorts of issues open.

**Comment by gchanan** [comment]
I agree with @colesbury -- I added "topic: determinism" in case we want to find related issues in the future, but addressing this seems unlikely so I'm going to close.
